pkg/operator/k8sutil/crud.go

```diff
@@ -0,0 +1,207 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+// DeleteDeployment makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
+func DeleteDeployment(clientset kubernetes.Interface, namespace, name string) error {
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().Deployments(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().Deployments(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "deployment", deleteAction, getAction)
+}
+
+// DeleteReplicaSet makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
+func DeleteReplicaSet(clientset kubernetes.Interface, namespace, name string) error {
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().ReplicaSets(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().ReplicaSets(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "replicaset", deleteAction, getAction)
+}
+
+// DeleteDaemonset makes a best effort at deleting a daemonset and its pods, then waits for them to be deleted
+func DeleteDaemonset(clientset kubernetes.Interface, namespace, name string) error {
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().DaemonSets(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().DaemonSets(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "daemonset", deleteAction, getAction)
+}
+
+// UpdateService updates a service, then waits for it to be updated. If the service does not exist,
+// this is considered an error condition.
+func UpdateService(clientset kubernetes.Interface, namespace string, serviceDefinition *v1.Service) error {
+	name := serviceDefinition.Name
+
+	existing, err := clientset.CoreV1().Services(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("could not get existing service %s in order to update. %+v", name, err)
+	}
+	// ClusterIP is immutable for k8s services and cannot be left empty in k8s v1 API
+	serviceDefinition.Spec.ClusterIP = existing.Spec.ClusterIP
+	// ResourceVersion required to update services in k8s v1 API to prevent race conditions
+	serviceDefinition.ResourceVersion = existing.ResourceVersion
+
+	// K8s resource definitions have a creation timestamp of the zero value, but it *is* technically
+	// defined, causing DeepDerivative to see that as an inequality.
+	serviceDefinition.CreationTimestamp = existing.CreationTimestamp
+	// If the service is already up-to-date on 'update', it doesn't get a new resource version,
+	// causing update to appear like a failure.
+	if k8sequality.Semantic.DeepDerivative(serviceDefinition, existing) {
```

Comment:
If we just call `Update()` on the service, won't it already determine equivalency for the update?